compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts

```diff
@@ -952,40 +812,52 @@ function mergeAbstractValues(
   return {kind, reason, context};
 }
 
+type Continuation =
+  | {
+      kind: 'initialize';
+      valueKind: AbstractValue;
+      effect: {kind: Effect; reason: ValueReason} | null;
+      lvalueEffect?: Effect;
+    }
+  | {kind: 'funeffects'};
+
 /*
  * Iterates over the given @param block, defining variables and
  * recording references on the @param state according to JS semantics.
  */
 function inferBlock(
   env: Environment,
-  functionEffects: Array<FunctionEffect>,
   state: InferenceState,
   block: BasicBlock,
+  functionEffects: Array<FunctionEffect>,
 ): void {
   for (const phi of block.phis) {
     state.inferPhi(phi);
   }
 
   for (const instr of block.instructions) {
     const instrValue = instr.value;
-    let effect: {kind: Effect; reason: ValueReason} | null = null;
-    let lvalueEffect = Effect.ConditionallyMutate;
-    let valueKind: AbstractValue;
+    const defaultLvalueEffect = Effect.ConditionallyMutate;
+    let continuation: Continuation;
+    const freezeActions: Array<FreezeAction> = [];
```

Comment:
As a side refactoring in this PR, I found the structure of the main IRE switch to be kind of confusing -- each case ending with `continue` if it's already performed all the actions it needs to, or with `break` if the case sets up the `effect` and `valueKind` variables to be handled in the fallthrough. This changes it to be explicitly controlled by a `continuation` variable (kind of a misnomer since it's not an actual function continuation, but \*shrug\*). All cases now break.